### ISSUE TITLE
weird detail field in 500 Internal Error messages 

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#843]
Testing MongoDB primary handover the following response is get:

```
{
  "contextResponses" : [
    {
      "contextElement" : {
        "type" : "device",
        "isPattern" : "false",
        "id" : "modelgtw2mysql3.iotasset2"
      },
      "statusCode" : {
        "code" : "500",
        "reasonPhrase" : "Internal Server Error",
        "details" : "\&quot;"
      }
    }
  ]
}
```

The problem is not getting a 500 error (which is normal during the primary handover) but the weird content of the `details` field. It should provide more meaningful information.

**Comment by kzangeli** [comment]
Seems like a single quote ' ...  Strange ... 
:-)

**Comment by fgalan** [comment]
Ideally, we should find a way of reproducing this without involving the complex test setup used for testing MongoDB primary handover (ideally it should be reproducible using stand alone MongoDB instance).

**Comment by XavierVal** [comment]
I hope this info help you to catch it.
This is the log when this error was showed again today: 
time=2015-05-05T11:39:28.716CEST | lvl=ERROR | trans=1430815520-585-00000000003 | function=processContextElement | comp=Orion | msg=MongoCommonUpdate.cpp[1666]: Database Error ('{ _id.id: "device_id1", _id.type: "thing", _id.servicePath: /^\/$|^\/\/.*/ }', 'socket exception [SEND_ERROR] for 10.0.0.xx:27017')

**Comment by kzangeli** [comment]
The version of the broker in the test described by @XavierVal  is a little old.
Looking at a newer version, it seems the error comer from line 1696 of MongoCommonUpdate.cpp (the line with the LM_E():

    catch (const DBException &e)
    {
        mongoSemGive(__FUNCTION__, "query in EntitiesCollection (mongo db exception)");
        buildGeneralErrorResponse(ceP, NULL, responseP, SccReceiverInternalError,
                           std::string("collection: ") + getEntitiesCollectionName(tenant).c_str() +
                              " - query(): " + query.toString() +
                              " - exception: " + e.what());
        LM_E(("Database Error ('%s', '%s')", query.toString().c_str(), e.what()));
        return;
    }

I have a feeling that if a send-error like this to mongod occurs, we should try again ...

**Comment by fgalan** [comment]
As far as I understand, this issue is not about that mongo connection error happens or not, but that, assuming that have appended, we see `"details" : "\&quot;"` in the response, instead of an actual description of the error, such as `"details" : "socket exception [SEND_ERROR] for 10.0.0.xx:27017" or similar`.

Maybe the details is "null-ified" in some point from buildGeneralErrorResponse() function creates it to the moment in which it is rendered in the upper layers? @kzangeli , could you have a look to this?

**Comment by fgalan** [comment]
Somehow related (user report shows a similar failure pattern): http://stackoverflow.com/questions/30061042/when-i-make-a-updatecontext-orion-shuts-down

**Comment by kzangeli** [comment]
Any idea on how to reproduce this error?
Anybody?

**Comment by fgalan** [comment]
Maybe setting the CB-MongoDB environment under sustained load (script, JMeter, etc.) and, in a given moment, use iptable to manually block the MongoDB port? (Just guessing :)

**Comment by XavierVal** [comment]
I try to reproduce it with the same steps showed in the log just before the "Internal Server Error" with three bad requests but I had not luck ...
```
time=2015-05-05T10:45:20.611CEST | lvl=INFO | trans=N/A | function=main | comp=Orion | msg=contextBroker.cpp[1434]: Startup completed
time=2015-05-05T11:12:52.383CEST | lvl=INFO | trans=1430815520-585-00000000001 | function=connectionTreat | comp=Orion | msg=rest.cpp[768]: Starting transaction from 10.95.230.79:51174/v1/updateContext
time=2015-05-05T11:12:52.384CEST | lvl=WARNING | trans=1430815520-585-00000000001 | function=badVerbPostOnly | comp=Orion | msg=badVerbPostOnly.cpp[54]: Bad Input (bad verb for url '/v1/updateContext', method 'GET')
time=2015-05-05T11:12:52.384CEST | lvl=INFO | trans=1430815520-585-00000000001 | function=requestCompleted | comp=Orion | msg=rest.cpp[382]: Transaction ended
time=2015-05-05T11:12:54.089CEST | lvl=INFO | trans=1430815520-585-00000000002 | function=connectionTreat | comp=Orion | msg=rest.cpp[768]: Starting transaction from 10.95.230.79:51175/v1/updateContext
time=2015-05-05T11:12:54.089CEST | lvl=WARNING | trans=1430815520-585-00000000002 | function=badVerbPostOnly | comp=Orion | msg=badVerbPostOnly.cpp[54]: Bad Input (bad verb for url '/v1/updateContext', method 'GET')
time=2015-05-05T11:12:54.089CEST | lvl=INFO | trans=1430815520-585-00000000002 | function=requestCompleted | comp=Orion | msg=rest.cpp[382]: Transaction ended
time=2015-05-05T11:39:28.714CEST | lvl=INFO | trans=1430815520-585-00000000003 | function=connectionTreat | comp=Orion | msg=rest.cpp[768]: Starting transaction from 10.95.230.79:62562/v1/updateContext
```

**Comment by mcarracedo** [comment]
Hi all,

I think I've found how to reproduce it. It seems to occur when fiware-service is empty ...

I paste my tests logs.

DEBUG >>>>>>>>>>>>>      Data sent:             >>>>>>>>>>>>>
        > Comp: CB
        > Url: http://192.168.22.137:1026/v1/updateContext
        > Method: post
        > Headers: {'Fiware-Service': u'', 'content-type': 'application/json', 'Fiware-Servicepath': u'/electricidad', 'Accept': 'application/json'}
        > Payload sent: '{"contextElements": [{"attributes": [{"type": "centigrade", "name": "temperature", "value": "2"}], "type": "entity", "id": "entity1071", "isPattern": "f
alse"}], "updateAction": "APPEND"}'

DEBUG <<<<<<<<<<<<<<     Data responded:        <<<<<<<<<<<<<<
        < Response code: 200
        < Headers: {'date': 'Tue, 26 May 2015 13:06:17 GMT', 'content-length': '458', 'content-type': 'application/json', 'connection': 'keep-alive', 'server': 'nginx/1.4.7'}
        < Payload received:
 {
  "contextResponses" : [
    {
      "contextElement" : {
        "type" : "entity",
        "isPattern" : "false",
        "id" : "entity1071",
        "attributes" : [
          {
            "name" : "temperature",
            "type" : "centigrade",
            "value" : ""
          }
        ]
      },
      "statusCode" : {
        "code" : "500",
        "reasonPhrase" : "Internal Server Error",
        "details" : "\&quot;"
      }
    }
  ]
}

**Comment by fgalan** [comment]
I'm trying to reproduce the problem with curl. In particular I'm using the following (note the `;` in `--header 'Fiware-Service;'`, which is the way of commanding curl to send an empty header):

```
(curl localhost:1026/v1/updateContext -s -S -vvvv --header 'Fiware-Service;' --header 'Fiware-Servicepath: /electricidad' --header 'Content-Type: application/json' --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
{
    "contextElements": [
        {
            "type": "entity",
            "isPattern": "false",
            "id": "entity1071",
            "attributes": [
            {
                "name": "temperature",
                "type": "centigrade",
                "value": "2"
            }
            ]
        }
    ],
    "updateAction": "APPEND"
}
EOF
```

However, the result seems to be correct:

```
* About to connect() to localhost port 1026 (#0)
*   Trying ::1...
* connected
* Connected to localhost (::1) port 1026 (#0)
> POST /v1/updateContext HTTP/1.1
> User-Agent: curl/7.26.0
> Host: localhost:1026
> Fiware-Service:
> Fiware-Servicepath: /electricidad
> Content-Type: application/json
> Accept: application/json
> Content-Length: 340
> 
} [data not shown]
* upload completely sent off: 340 out of 340 bytes
* additional stuff not fine transfer.c:1037: 0 0
* HTTP 1.1 or later with persistent connection, pipelining supported
< HTTP/1.1 200 OK
< Content-Length: 408
< Content-Type: application/json
< Date: Wed, 27 May 2015 07:58:12 GMT
< 
{ [data not shown]
* Connection #0 to host localhost left intact
* Closing connection #0
{
    "contextResponses": [
        {
            "contextElement": {
                "attributes": [
                    {
                        "name": "temperature", 
                        "type": "centigrade", 
                        "value": ""
                    }
                ], 
                "id": "entity1071", 
                "isPattern": "false", 
                "type": "entity"
            }, 
            "statusCode": {
                "code": "200", 
                "reasonPhrase": "OK"
            }
        }
    ]
}
```

Thus, two questions @mcarracedo:

* Using your testing tool, are your able to reproduce the problem each time? Or it happens randomly?
* In the former case, could you show me (maybe offline) the exact message we have on the wire (wireshark, etc.) so I can try to mimic it with curl?

**Comment by fgalan** [comment]
Fortunatelly We have found a way of reproducing this by change reseraching on a different bug.

Fixed in PR #1145
